File: daily_python/3337_Total_Characters_in_String_After_Transformations_II.py
import os
import sys
from typing import List

class Solution:
    def mul(self, s: List[List[int]], t: List[List[int]]) -> List[List[int]]:
        mod = 10 ** 9 + 7
        m = len(s)
        n = len(s[0])
        k = len(t[0])
        if len(t) != n:
            return []
        ans = [[0] * k for _ in range(m)]
        for i in range(m):
            for j in range(k):
                for l in range(n):
                    ans[i][j] = (ans[i][j] + (s[i][l] * t[l][j])) % mod
        return ans
    
    def lengthAfterTransformations(self, s: str, t: int, nums: List[int]) -> int:
        mod = 10 ** 9 + 7
        # [1x26] * ([26 * 26] ^ t) = [1 * 26]
        init = [[0] * 26]
        for c in s:
            init[0][ord(c) - ord('a')] += 1
        trans = [[0] * 26 for _ in range(26)]
        for i in range(26):
            for j in range(1, nums[i] + 1):
                trans[i][(i + j) % 26] = 1
        res = [[0] * 26 for _ in range(26)]
        for i in range(26):
            res[i][i] = 1
        while t:
            if t % 2 == 1:
                res = self.mul(res, trans)
            trans = self.mul(trans, trans)
            t //= 2
        ending = self.mul(init, res)[0]
        ans = 0
        for cnt in ending:
            ans = (ans + cnt) % mod
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            s = lines[0].split('=')[1].strip()[1:-2]
            t = int(lines[1].split('=')[1].strip()[:-1])
            nums = list(map(int, lines[2].split('=')[1].strip()[1:-2].split(',')))
            dataset.append((s, t, nums))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for s, t, nums in dataset:
        results.append(solution.lengthAfterTransformations(s, t, nums))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2025/3337_Total_Characters_in_String_After_Transformations_II.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    int mod = 1e9 + 7;
    vector<vector<long long>> mul(vector<vector<long long>>& a, vector<vector<long long>>& b){
        int m = a.size();
        int n = a[0].size();
        if (n != b.size()) return {{}};
        int k = b[0].size();
        vector<vector<long long>> ans(m, vector<long long>(k, 0));
        for (int i = 0; i < m; i++){
            for (int j = 0; j < k; j++){
                long long res = 0;
                for (int l = 0; l < n; l++){
                    res += ((a[i][l] * b[l][j]) % mod);
                    res %= mod;
                }
                ans[i][j] = res;
            }
        }
        return ans;
    }
    int lengthAfterTransformations(string s, int t, vector<int>& nums) {
        vector<vector<long long>> init(26, vector<long long>(1, 0));
        for (char c : s) init[c-'a'][0]++;
        vector<vector<long long>> transformer(26, vector<long long>(26, 0));
        for (int i = 0; i < 26; i++){
            for (int j = 1; j <= nums[i]; j++){
                transformer[(i + j) % 26][i]++;
            }
        }
        vector<vector<long long>> res(26, vector<long long>(26, 0));
        for (int i = 0; i < 26; i++) res[i][i] = 1;
        while(t){
            if (t % 2 == 1) res = mul(res, transformer);
            transformer = mul(transformer, transformer);
            t /= 2;
        }
        vector<vector<long long>> final = mul(res, init);
        long long ans = 0;
        for (int i = 0; i < 26; i++){
            ans = (ans + final[i][0]) % mod;
        }
        return ans;
    }
};

int main(){
    string s = "abcyy";
    int t = 2;
    vector<int> nums = {1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2};
    // string s = "azbk";
    // int t = 1;
    // vector<int> nums = {2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2};
    Solution S;
    cout << S.lengthAfterTransformations(s, t, nums) << endl;
    return 0;
}